pkg/proxy/proxy.go

```diff
@@ -286,12 +288,44 @@ func uploadToBlobStorage(ctx context.Context, cacheID int) (*DockerGHAUploadCach
 			}
 		}
 
+	case ProviderGCS:
+		if cacheEntry.BackendReserveResponse.GCS.ShortLivedToken == nil {
+			return nil, fmt.Errorf("no short lived token found")
+		}
+
+		creds := option.WithCredentialsJSON([]byte(cacheEntry.BackendReserveResponse.GCS.ShortLivedToken.AccessToken))
+		client, err := storage.NewClient(ctx, creds)
 		if err != nil {
-			return nil, fmt.Errorf("upload to S3 failed after %d attempts: %w", maxRetries, err)
+			return nil, fmt.Errorf("failed to create GCS client: %w", err)
 		}
 
-	} else {
+		defer client.Close()
+
+		bucket := client.Bucket(cacheEntry.BackendReserveResponse.GCS.BucketName)
+		object := bucket.Object(cacheEntry.BackendReserveResponse.GCS.CacheKey)
+
+		wc := object.NewWriter(ctx)
+
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			_, err = wc.Write(finalBuffer.Bytes())
+			if err == nil {
+				err = wc.Close()
+				if err == nil {
+					return &DockerGHAUploadCacheResponse{}, nil
+				}
+			}
+
+			if attempt < maxRetries-1 {
+				fmt.Printf("Retrying upload... attempt %d/%d, error: %v\n", attempt+1, maxRetries, err)
+				time.Sleep(time.Duration(1<<attempt) * time.Second)
```

Comment:
Nit: `time.Duration` wrap isn't needed here:

```suggestion
				time.Sleep((1<<attempt) * time.Second)
```

Reply:
Done.